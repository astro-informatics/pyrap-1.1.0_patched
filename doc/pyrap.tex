\section{Introduction}
Since long glish bindings to the \href{http://casacore.googlecode.com}{casacore} system have been in
place. Quite recently Python bindings have been created in the general
casapy framework using tools like CCMTools, Xerces, Xalan, and
IDL. Albeit very flexible, it is quite complicated and it is not
straightforward to build on other systems than RedHat and OS-X.

Therefore an attempt has been made to make a simpler Python binding
using \href{http://www.boost.org/libs/python/doc}{Boost.Python}.
This proved to be very easy and succesful.
The binding consists of two parts:
\begin{itemize}
\item Converters to translate objects between Python and C++.
\item Class wrappers to map a C++ class and its functions to Python.
\end{itemize}
The Python numarray and numpy (version 1.0 or higher) packages are
supported. At build time one can choose which ones should be used.

\section{Converters}
Boost.Python offers a nice way to convert objects to and from Python.
Ralf W. Grosse-Kunstleve {\tt<rwgk@yahoo.com>} of
\href{http://www.lbl.gov}{Lawrence Berkeley National Laboratory}
has built converters for standard STL containers. This has been
extended to convert to/from other objects.
\\The following C++ objects are currently supported:
\begin{itemize}
  \item scalars (bool, integer, real, complex)
  \item \texttt{std::string}
  \item \texttt{casa::String}
  \item \texttt{std::vector<T>}
  \item \texttt{casa::Vector<T>}
  \item \texttt{casa::IPosition} which defines an array's shape or index.
  \item \texttt{casa::Record} which is a \tt{dict}-like object.
  \item \texttt{casa::ValueHolder} which is a kind of \tt{any} object
    holding a scalar, casa::Array or casa::Record.
  \item exceptions (\texttt{casa::IterError} and \texttt{std::exception})
\end{itemize}
These C++ objects can usually be created from several types of Python
objects and are converted to a specific Python object.
\\Note that the all kinds of numpy (and numarray) objects are
handled by the converters. These are:
\begin{itemize}
\item An array is an N-dimensional (N>0) array of a given type
  (bool, int, float, complex, string).
\item A scalar array is a 0-dimensional array constructed like
  \tt{numpy.array(value)}. Such an array cannot be indexed in Python.
\item An array scalar is an element from an array. The converters
  treat them as normal scalars, which very few Python bindings do.
\end{itemize}
The C++ objects map in the following way to Python objects.
\begin{itemize}
\item
A vector or IPosition object is converted to a Python list.
\\It can be constructed from the following Python objects:
\begin{itemize}
  \item scalar
  \item list or tuple
  \item numarray scalar array or 1-dim array
  \item numpy scalar array or 1-dim array
\end{itemize}
Note that a list or tuple of arbitrary objects can be given. For
example, it is possible to get a \texttt{Vector<TableProxy>} from Python.
\item
A casa::Record is mapped to a Python dict.
\item
Every C++ exception is mapped to a Python \texttt{RunTimeError}
exception. However, \texttt{casa::IterError} is special and 
is mapped to an end-of-iteration exception (\texttt{StopIteration}) in Python.
\item
A casa::ValueHolder is a special \href{http://casacore.googlecode.com}{casacore} object that can hold a record or
a scalar value or N-dim array of many types (bool, numeric, string).
It is meant to conceal the actual type which is useful in functions
that can accept a variety of types (like \texttt{getcell}
in the table binding).
\\Converting a ValueHolder to Python creates the appropriate Python
scalar, array, or dict object. When converting from
Python to ValueHolder, the appropriate internal ValueHolder value
is constructed; a list, tuple, and array object are converted to an
\href{http://casacore.googlecode.com}{casacore} array in the ValueHolder.
\end{itemize}
It means there is no direct Array conversion to/from Python. A
ValueHolder object is always needed to do the conversion. Note that
this is a cheap operation, as it uses Array reference semantics.
ValueHolder has functions to convert between types, so one can get
out an Array with the required type.

\subsection{Array conversion to/from numpy and numarray}
\href{http://casacore.googlecode.com}{casacore} arrays are kept in Fortran-order, while Python arrays are kept
in C-order. It was felt that the Python interface should be as 
pythonic as possible. Therefore it was decided that the array axes are
reversed when converting to/from Python.
The values in an IPosition object (describing shape or
position) are also reversed when converting to/from Python.
\\Note that although numarray and numpy have Fortran-array provisions
by setting the appropriate internal strides,
they do not really support them. When adding, for instance, the scalar
value 0 to a Fortran-array, the  result is a transposed version of the
original (which can be a quite expensive operation). Also operating on
a Fortran-array is twice as slow as operating on a C-array.

A function binding could be such that shape information is passed via, say, a
\texttt{Record} and not via an \texttt{IPosition} object.
In that case its values
are not reversed automatically, so the programmer is
responsible for doing it.

An \href{http://casacore.googlecode.com}{casacore} array is returned to Python as an array object
containing a copy of the \href{http://casacore.googlecode.com}{casacore} array data. If pyrap has
been built with support for only one Python array package (numpy or numarray),
it is clear which array type is returned. If support for both
packages has been built in, by default an array of the imported
package is returned. If both or no array packages have been imported, a numpy
array is returned.
\\Note that there is no support for the old Numeric package.

An \href{http://casacore.googlecode.com}{casacore} array constructed from a Python array is regarded as a
temporary object. So if possible, the \href{http://casacore.googlecode.com}{casacore} array refers to the
Python array data to avoid a needless copy. This is not possible if
the element size in Python differs from \href{http://casacore.googlecode.com}{casacore}. It is also not
possible if the Python array is not contiguous (or not aligned or
byte swapped). In those cases a copy is made.

A few more numarray/numpy specific issues are dealt with:
\begin{itemize}
\item An empty N-dim \href{http://casacore.googlecode.com}{casacore} array (i.e. an array containing no elements) is
returned as an empty N-dim Python array. If the dimensionality is
zero, it is returned as an empty 1-dim array, to prevent numarray/numpy
from treating it as a scalar value.
\item In numarray \texttt{array()} results in \texttt{Py\_None}. This
is accepted by the converters as an empty 1-dim array.
\item Empty arrays can be constructed in Python using empty lists. For
example, \texttt{array([[]])} results in an empty 2-dim array. The
converters accept such empty N-dim Python arrays. The type of an empty
array is set to Int by numarray and to Double by numpy.
\item Because the type of an empty Python array cannot easily be set,
the converters can convert an empty integer or real array to any type.
\item The converters accept a numpy string array. However, it is
returned to Python as the special \texttt{dict} object described
above.
\item Numpy array scalars (elements in an array) are treated by the
converters as ordinary scalar values.
\end{itemize}

\section{Class wrappers}
Usually a binding to an existing Proxy class is made, for example
\texttt{TableProxy}, which should be the same class used in the
glish-binding. 
For a simple binding, only some simple C++ code has to be written in
pyrap\_xx/src/pyxx.cc, where XX is the name of the package/class.
\begin{verbatim}
// Include files for converters being used.
#include <pyrap/Converters/PycExcp.h>
#include <pyrap/Converters/PycBasicData.h>
#include <pyrap/Converters/PycRecord.h>
// Include file for boost python.
#include <boost/python.hpp>

using namespace boost::python;

namespace casa { namespace pyrap {
  void wrap_xx()
  {
    // Define the class; "xx" is the class name in Python.
    class_<XX> ("xx")
      // Define the constructor.
      // Multiple constructors can be defined.
      // They have to have different number of arguments.
      .def (init<>())
      // Add a .def line for each function to be wrapped.
      // An arg line should be added for each argument giving
      // its name and possibly default value.
      .def ("func1", &XX::func1,
            (boost::python::arg("arg1"),
             boost::python::arg("arg2")=0))
    ;
  }
}}

BOOST_PYTHON_MODULE(_xx)
{
  // Register the conversion functions.
  casa::pyrap::register_convert_excp();
  casa::pyrap::register_convert_basicdata();
  casa::pyrap::register_convert_casa_record();
  // Initialize the wrapping.
  casa::pyrap::wrap_xx();
}
\end{verbatim}
Python requires for each package a file \texttt{\_\_init\_\_.py},
so such an empty file should be created as well.

\subsection{More complicated wrappers}
Sometimes a C++ function cannot be wrapped directly, because the argument
order needs to be changed or because some extra Python checks are
necessary.
In such a case the class needs to be implemented in Python itself.
\\The C++ wrapped class name needs to get a different name,
usually by preceeding it with an underscore like:
\begin{verbatim}
    class_<XX> ("_xx")
\end{verbatim}
The Python class should be derived from it and implement the
constructor by calling the constructor of \_xx.
\begin{verbatim}
class xx(_xx):
    def __init__(self):
        _xx.__init__(self)
\end{verbatim}
Now \texttt{xx} inherits all functions from \texttt{\_xx}.
The required function can be written in Python like
\begin{verbatim}
    def func1 (self, arg1, arg2):
        return self._func1 (arg2, arg1);
\end{verbatim}
Note that in the wrapper the function name also needs to be
preceeded by an underscore to make it different.

\subsection{Combining multiple classes}
Sometimes one wants to combine multiple classes in a package. A
example is package \texttt{pyrap\_tables} which contains the classes
\texttt{table}, \texttt{tablecolumn}, \texttt{tablerow},
\texttt{tableiter}, and \texttt{tableindex}. One is referred to the
code of this package to see how
to do it.

\section{Python specifics}
Besides an array being in C-order, there are a few more Python
specific issues.
\begin{itemize}
\item Indexing starts at 0 (vs. 1 in glish).
\item The end value in a range like \texttt{[10:20]} is exclusive
(vs. inclusive in glish). Furthermore Python supports a step and
reversed ranges.
\item Where useful, the function \texttt{\_\_str\_\_} should be added
  giving the name of the object. This function is used when printing
  an object.
\item Where useful, the functions \texttt{\_\_len\_\_},
\texttt{\_\_setitem\_\_(index, value)}, and 
\texttt{\_\_getitem\_\_(index)} should
be added to make it possible that a user indexes an object directly like
\texttt{tabcol[i]} or \texttt{tabcol[start:stop:step]}.
\item When these functions are added, Python supports iteration in an
object. Explicit iteration can also be done by adding the functions
\texttt{\_\_iter\_\_} and \texttt{next}. At the end \texttt{next}
should raise the Python \texttt{StopIteration} exception
(or throw \texttt{casa::IterError}
when implemented in C++) to stop the iteration.
\end{itemize}
