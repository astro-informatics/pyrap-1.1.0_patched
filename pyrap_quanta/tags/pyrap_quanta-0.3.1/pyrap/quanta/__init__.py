from quantity import quantity, is_quantity
from _quanta import *
constants = constants()
del Quantity, QuantVec, from_string, from_dict_v
