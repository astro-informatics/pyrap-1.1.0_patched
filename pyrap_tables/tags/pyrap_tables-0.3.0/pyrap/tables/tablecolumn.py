# tablecolumn.py: Python tablecolumn functions
# Copyright (C) 2006
# Associated Universities, Inc. Washington DC, USA.
#
# This library is free software; you can redistribute it and/or modify it
# under the terms of the GNU Library General Public License as published by
# the Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
# License for more details.
#
# You should have received a copy of the GNU Library General Public License
# along with this library; if not, write to the Free Software Foundation,
# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
#
# Correspondence concerning AIPS++ should be addressed as follows:
#        Internet email: aips2-request@nrao.edu.
#        Postal address: AIPS++ Project Office
#                        National Radio Astronomy Observatory
#                        520 Edgemont Road
#                        Charlottesville, VA 22903-2475 USA
#
# $Id: tablecolumn.py,v 1.9 2007/08/28 07:22:18 gvandiep Exp $

class tablecolumn:
    """The Python interface to a column in a Casacore table.

    The `tablecolumn` class is a convenience class to access data in a
    table column. All functionality provided in this class is available in
    :class:`table`, but `tablecolumn` is more convenient to use because the
    column name does not have to be given over and over again.

    For example::

      t = table('3C343.MS')
      tc = tablecolumn(t, 'DATA')
      # tc = t.col('DATA')           # another way to construct a tablecolumn
      tc.getcell(0)                  # get data from cell 0

    As can be seen in the example :func:`table.col` offers a slightly more
    convenient way to create a `tablecolumn` object.

    A `tablecolumn` can be indexed using Python's [] operator. Negative start,
    end, and stride is possible. For example::

      tc[0]               # get cell 0
      tc[:5]              # get cell 0,1,2,3,4
      tc[-5,-1,]          # get last 4 cells
      tc[-1,-5,-1]        # get last 4 cells in reversed order
      tc[1] = tr[0]       # put value of cell 0 into cell 1

    """

    def __init__(self, table, columnname):
        if not columnname in table.colnames():
            raise RuntimeError("Column " + columnname + " does not exist in table " + table.name());
        self._table  = table;
        self._column = columnname;

    def name (self):
        """Get the name of the column."""
        return self._column;

    def table (self):
        """Get the table object this column belongs to."""
        return self._table;

    def isscalar (self):
        """Tell if the column contains scalar values."""
        return self._table.isscalarcol (self._column);

    def isvar (self):
        """Tell if the column holds variable shaped arrays."""
        return self._table.isvarcol (self._column);

    def datatype (self):
        """Get the data type of the column.
        (see :func:`table.coldatatype`)"""
        return self._table.coldatatype (self._column);

    def arraytype (self):
        """Get the array type of a column holding arrays.
        (see :func:`table.colarraytype`)"""
        return self._table.colarraytype (self._column);

    def nrows (self):
        """Get number of cells in the column."""
        return self._table.nrows();

    def getshapestring (self, startrow=1, nrow=-1, rowincr=1):
        """Get the shapes of all cells in the column in string format.
        (see :func:`table.getcolshapestring`)"""
        return self._table.getcolshapestring (self._column,
                                              startrow, nrow, rowincr);

    def iscelldefined (self, rownr):
        """Tell if a column cell contains a value.
        (see :func:`table.iscelldefined`)"""
        return self._table.iscelldefined (self._column, rownr);

    def getcell (self, rownr):
        """Get data from a column cell.
        (see :func:`table.getcell`)"""
        return self._table.getcell (self._column, rownr);

    def getcellslice (self, rownr, blc, trc, inc=[]):
        """Get a slice from a column cell holding an array.
        (see :func:`table.getcellslice`)"""
        return self._table.getcellslice (self._column, rownr, blc, trc, inc);

    def getcol (self, startrow=0, nrow=-1, rowincr=1):
        """Get the contents of the column or part of it.
        (see :func:`table.getcol`)"""
        return self._table.getcol (self._column, startrow, nrow, rowincr);

    def getvarcol (self, startrow=0, nrow=-1, rowincr=1):
        """Get the contents of the column or part of it.
        (see :func:`table.getvarcol`)"""
        return self._table.getvarcol (self._column, startrow, nrow, rowincr);

    def getcolslice (self, blc, trc, inc=[], startrow=0, nrow=-1, rowincr=1):
        """Get a slice from a table column holding arrays.
        (see :func:`table.getcolslice`)"""
        return self._table.getcolslice (self._column, blc, trc, inc, startrow, nrow, rowincr);

    def putcell (self, rownr, value):
        """Put a value into one or more table cells.
        (see :func:`table.putcell`)"""
        return self._table.putcell (self._column, rownr, value);

    def putcellslice (self, rownr, value, blc, trc, inc=[]):
        """Put into a slice of a table cell holding an array.
        (see :func:`table.putcellslice`)"""
        return self._table.putcellslice (self._column, rownr, value, blc, trc, inc);

    def putcol (self, value, startrow=0, nrow=-1, rowincr=1):
        """Put an entire column or part of it.
        (see :func:`table.putcol`)"""
        return self._table.putcol (self._column, value, startrow, nrow, rowincr);

    def putvarcol (self, value, startrow=0, nrow=-1, rowincr=1):
        """Put an entire column or part of it.
        (see :func:`table.putvarcol`)"""
        return self._table.putvarcol (self._column, value, startrow, nrow, rowincr);

    def putcolslice (self, value, blc, trc, inc=[], startrow=0, nrow=-1, rowincr=1):
        """Put into a slice in a table column holding arrays.
        (see :func:`table.putcolslice`)"""
        return self._table.putcolslice (self._column, value, blc, trc, inc, startrow, nrow, rowincr);

    def keywordnames (self):
        """Get the names of all keywords of the column."""
        return self._table.colkeywordnames (self._column);

    def fieldnames (self, keyword=''):
        """Get the names of the fields in a column keyword value.
        (see :func:`table.colfieldnames`)"""
        return  self._table.colfieldnames (self._column, keyword);

    def getkeyword (self, keyword):
        """Get the value of a column keyword.
        (see :func:`table.getcolkeyword`)"""
        return  self._table.getcolkeyword (self._column, keyword);

    def getkeywords (self):
        """Get the value of all keywords of the column.
        (see :func:`table.getcolkeywords`)"""
        return self._table.getcolkeywords (self._column);

    def putkeyword (self, keyword, value, makesubrecord=False):
        """Put the value of a column keyword.
        (see :func:`table.putcolkeyword`)"""
        return self._table.putcolkeyword (self._column, keyword, value, makesubrecord);

    def putkeywords (self, value):
        """Put the value of multiple table keywords.
        (see :func:`table.putcolkeywords`)"""
        return self._table.putcolkeywords (self._column, value);

    def removekeyword (self, keyword):
        """Remove a column keyword.
        (see :func:`table.removecolkeyword`)"""
        return self._table.removecolkeyword (self._column, keyword);

    def getdesc (self):
        """Get the description of the column.
        (see :func:`table.getcoldesc`)"""
        return self._table.getcoldesc (self._column);

    def getdminfo (self):
        """Get data manager info of the column.
        (see :func:`table.getdminfo`)"""
        return self._table.getdminfo (self._column);

    def iter (self, order='', sort=True):
        """Return a :class:`tableiter` object on this column."""
        from tables import tableiter;
        return tableiter (self._table, [self._column], order, sort);

    def index (self, sort=True):
        """Return a :class:`tableindex` object on this column."""
        from tables import tableindex;
        return tableindex (self._table, [self._column], sort);

    def __len__ (self):
        return self._table.nrows();

    def __getitem__ (self, key):
        sei = self.checkkey (key, self._table.nrows());
        if len(sei) == 1:
            # A single row.
            return self.getcell (sei[0]);
        # Handle row by row and store values in a list.
        result = [];
        rownr  = sei[0];
        inx    = 0;
        while inx < sei[1]:
            result.append (self.getcell (rownr));
            rownr += sei[2];
            inx   += 1;
        return result;

    def __setitem__ (self, key, value):
        sei = self.checkkey (key, self._table.nrows());
        if len(sei) == 1:
            # A single row.
            return self.putcell (sei[0], value);
        # Handle row by row.
        rownr = sei[0];
        inx   = 0;
        if not (isinstance(value,list) or isinstance(value,tuple)):
            # The same value is put in all rows.
            while inx < sei[1]:
                self.putcell (rownr, value);
                rownr += sei[2];
                inx   += 1;
        else:
            # Each row has its own value.
            if len(value) != sei[1]:
                raise RuntimeError("tablecolumn slice length differs from value length")
            for val in value:
                self.putcell (rownr, val);
                rownr += sei[2];
        return True;
        
    def checkkey (self, key, nrows):
        if not isinstance(key, slice):
            # A single index (possibly negative, thus from the end).
            if key < 0:
                key += nrows;
            if key < 0  or  key >= nrows:
                raise IndexError("tablecolumn index out of range");
            return [key];
        # Given as start:stop:step where each part is optional and can
        # be negative.
        incr = 1;
        if key.step != None:
            incr = key.step;
            if incr == 0:
                raise RunTimeError("tablecolumn slice step cannot be zero");
        strow  = 0;
        endrow = nrows;
        if incr < 0:
            strow  = nrows-1;
            endrow = -1;
        if key.start != None:
            strow = key.start;
            if strow < 0:
                strow += nrows;
            strow = min(max(strow,0), nrows-1);
        if key.stop != None:
            endrow = key.stop;
            if endrow < 0:
                endrow += nrows;
            endrow = min(max(endrow,-1), nrows);
        if incr > 0:
            nrow = int((endrow - strow + incr - 1) / incr);
        else:
            nrow = int((strow - endrow - incr - 1) / -incr);
        nrow = max(0, nrow);
        return [strow,nrow,incr];
