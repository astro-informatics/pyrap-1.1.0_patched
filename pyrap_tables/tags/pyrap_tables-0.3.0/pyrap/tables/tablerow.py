# tablerow.py: Python tablerow functions
# Copyright (C) 2006
# Associated Universities, Inc. Washington DC, USA.
#
# This library is free software; you can redistribute it and/or modify it
# under the terms of the GNU Library General Public License as published by
# the Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
# License for more details.
#
# You should have received a copy of the GNU Library General Public License
# along with this library; if not, write to the Free Software Foundation,
# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
#
# Correspondence concerning AIPS++ should be addressed as follows:
#        Internet email: aips2-request@nrao.edu.
#        Postal address: AIPS++ Project Office
#                        National Radio Astronomy Observatory
#                        520 Edgemont Road
#                        Charlottesville, VA 22903-2475 USA
#
# $Id: tablerow.py,v 1.6 2007/08/28 07:22:18 gvandiep Exp $

# Make interface to class TableRowProxy available.
from _tables import TableRow

# A normal tablerow object keeps a reference to a table object to be able
# to know the actual number of rows.
# However, a mutual dependency is created when doing that for the tablerow
# object inside the table object.
# Therefore an intermediate _tablerow exists to be used in class table.

class _tablerow(TableRow):
    def __init__(self, table, columnnames, exclude=False):
        TableRow.__init__ (self, table, columnnames, exclude);
    
    def iswritable(self):
        """Tell if all columns in the row object are writable."""
        return self._iswritable()

    def get (self, rownr):
        """Get the contents of the given row."""
        return self._get (rownr)

    def put (self, rownr, value, matchingfields=True):
        """Put the values into the given row.

        The value should be a dict (as returned by method :func:`get`.
        The names of the fields in the dict should match the names of the
        columns used in the `tablerow` object.

        `matchingfields=True` means that the value may contain more fields
        and only fields matching a column name will be used.

        """
        self._put (rownr, value, matchingfields)

    def _getitem (self, key, nrows):
        sei = self.checkkey (key, nrows);
        rownr = sei[0];
        if len(sei) == 1:
            return self.get (rownr);
        result = [];
        inx = 0;
        while inx < sei[1]:
            result.append (self.get (rownr));
            rownr += sei[2];
            inx   += 1;
        return result;
    
    def _setitem (self, key, value, nrows):
        sei = self.checkkey (key, nrows);
        rownr = sei[0];
        if len(sei) == 1:
            return self.put (rownr, value);
        if isinstance(value, dict):
            # The same value is put in all rows.
            inx = 0;
            while inx < sei[1]:
                self.put (rownr, value, True);
                rownr += sei[2];
                inx   += 1;
        else:
            # Each row has its own value.
            if len(value) != sei[1]:
                raise RuntimeError("tablerow slice length differs from value length")
            for val in value:
                self.put (rownr, val, True);
                rownr += sei[2];

    def checkkey (self, key, nrows):
        if not isinstance(key, slice):
            if key < 0:
                key += nrows;
            if key < 0  or  key >= nrows:
                raise IndexError("tablerow index out of range");
            return [key];
        incr = 1;
        if key.step != None:
            incr = key.step;
            if incr == 0:
                raise RunTimeError("tablerow slice step cannot be zero");
        strow  = 0;
        endrow = nrows;
        if incr < 0:
            strow  = nrows-1;
            endrow = -1;
        if key.start != None:
            strow = key.start;
            if strow < 0:
                strow += nrows;
            strow = min(max(strow,0), nrows-1);
        if key.stop != None:
            endrow = key.stop;
            if endrow < 0:
                endrow += nrows;
            endrow = min(max(endrow,-1), nrows);
        if incr > 0:
            nrow = int((endrow - strow + incr - 1) / incr);
        else:
            nrow = int((strow - endrow - incr - 1) / -incr);
        nrow = max(0, nrow);
        return [strow,nrow,incr];



class tablerow(_tablerow):
    """The Python interface to Casacore table rows.

    A table row is a record (dict) containing the values of a single row for
    one or more columns in a table. In constructing the `tablerow` object, one
    can specify which columns are to be included or excluded.
    By default all columns will be used, but if the table is writable,
    only writable columns will be used.

    A `tablerow` object can easily be constructed using :func:`table.row`.

    One or more rows can be read or written using the standard python indexing
    syntax where (negative) strides are possible.
    For example:

      t = table ('3c343.MS')
      tr = t.row (['ANTENNA1', 'ANTENNA2', 'ARRAY_ID'])
      tr[0]               # get row 0
      tr[:5]              # get row 0,1,2,3,4
      tr[-5,-1,]          # get last 4 rows
      tr[-1,-5,-1]        # get last 4 rows in reversed order
      tr[1] = tr[0]       # put values of row 0 into row 1

    Note that the last line will fail because the table is opened readonly.
    The argument `readonly=False` is needed in the table constructor to make
    it work.

    """
    def __init__(self, table, columnnames=[], exclude=False):
        _tablerow.__init__ (self, table, columnnames, exclude);
        self._table = table;

    def __len__ (self):
        return self._table.nrows();

    def __getitem__ (self, key):
        return self._getitem (key, self._table.nrows());

    def __setitem__ (self, key, value):
        return self._setitem (key, value, self._table.nrows());
