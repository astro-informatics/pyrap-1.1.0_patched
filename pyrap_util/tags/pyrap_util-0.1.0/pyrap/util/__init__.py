"""
Utilities for pyrap modules.
"""
from substitute import substitute, getlocals, getvariable
