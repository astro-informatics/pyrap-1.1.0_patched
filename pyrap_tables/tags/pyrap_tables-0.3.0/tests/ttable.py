#!/usr/bin/env python

from pyrap.tables import *
import numpy

# Make some columns (5 scalars and an array)
c1 = makescacoldesc ("coli", 0)
c2 = makescacoldesc ("cold", 0.)
c3 = makescacoldesc ("cols", "")
c4 = makescacoldesc ("colb", True)
c5 = makescacoldesc ("colc", 0.+0j)
c6 = makearrcoldesc ("colarr", 0.)
t = table ("ttable.py_tmp.tab1", maketabdesc((c1,c2,c3,c4,c5,c6)), ack=False)
# Print some info
print t.nrows()
print t.colnames()
print t.getkeywords()
print t.info()
t.addreadmeline ("test table run")
t.putinfo({'type':'test'})
print t.info()
# Add rows and put data
t.addrows (2)
print t.nrows()
t.putcol ("coli", (1,2))
t.putcol ("cold", t.getcol('coli')+3)
# Create a subset
t1 = t.query('coli == 3')
print len(t1)
t1 = taql('select coli,cold from $t where coli>0 order by coli desc')
print t1.colnames()
print t1.getcol('cold')
t1.close()

# Add some columns
# A scalar with the IncrementalStMan storage manager
t.addcols (maketabdesc(makescacoldesc("coli2",0)),
           dminfo={'TYPE':"IncrementalStMan", 'NAME':"ism1", 'SPEC':{}})
print t.colnames()
# An array with the StandardStMan
t.addcols (maketabdesc(makearrcoldesc("colarrssm","")))
# An array with the TiledShapeStMan
t.addcols (maketabdesc(makearrcoldesc("colarrtsm",0.+0j, ndim=2)),
           dminfo={'TYPE':"TiledShapeStMan", 'NAME':"tsm1", 'SPEC':{}})
print t.getdminfo()
coldmi = t.getdminfo('colarrtsm')
print t.getcoldesc('colarrtsm')
coldmi["NAME"] = 'tsm2'
t.addcols (maketabdesc(makearrcoldesc("colarrtsm2",0., ndim=2)), coldmi)
print t.getdminfo('colarrtsm2')
t.removecols('colarrtsm2')
# Write some data.
t.putcell('colarrtsm', 0, numpy.array([[1,2,3],[4,5,6]]))
t.putcell('colarrtsm', 1, t.getcell('colarrtsm',0)+10)
print t.getcol('colarrtsm')
print t.getcellslice('colarrtsm', 0, [1,1], [1,2])

# Do keyword handling
t.putkeyword ('key1', "keyval")
t.putkeyword ('keyrec', {'skey1':1, 'skey2': 3.})
print t.getkeywords()
print t.getkeyword('key1')
print t.keywordnames()
print t.fieldnames()
print t.fieldnames('keyrec')
print t.getcolkeywords('coli')

# Delete some columns.
t.removecols(['colarr','colarrssm','cols','colb','colc','coli2'])
print t.colnames()
# Print table row
tr = t.row(['coli','cold','colarrtsm'])
print tr[0]
# Update a few fields in the row
tr[0] = {'coli':10, 'cold':14}
# Rename a column
t.renamecol ('colarrtsm', 'colarr')
print t[0]

# Same TaQL calculations
print t.calc("(1 km)cm")
print t.calc("coli+1")

# Add some more data.
t.addrows(20);
for i in range(2,22):
    t.putcell ('coli', i, i/2)
# Table iteration
for iter in t.iter('coli'):
    print iter.getcol('coli'), iter.rownumbers(t)
# Table column
tc = tablecolumn(t,'coli');
tc[6] += 20
print tc[18:4:-2]
print tc[0:]
# Table index
ti = t.index('coli')
print ti.isunique(), ti.colnames()
print ti.rownrs(23)
print ti.rownrs(20)
print ti.rownrs(2)
print ti.rownrs(2,7)                   # include borders
print ti.rownrs(2,7,False,False)       # exclude borders
print ti[2:7]                          # exclude end
