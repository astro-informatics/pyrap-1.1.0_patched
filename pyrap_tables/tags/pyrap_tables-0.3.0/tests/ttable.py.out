0
['cols', 'colc', 'coli', 'cold', 'colb', 'colarr']
{}
{'readme': '', 'subType': '', 'type': ''}
{'readme': 'test table run\n', 'subType': '', 'type': 'test'}
2
0
['coli', 'cold']
[ 5.  4.]
['cols', 'colc', 'coli', 'cold', 'colb', 'colarr', 'coli2']
{'*1': {'TYPE': 'StandardStMan', 'NAME': 'StandardStMan', 'COLUMNS': ['colarr', 'colarrssm', 'colb', 'colc', 'cold', 'coli', 'cols'], 'SPEC': {'CACHESIZE': 2, 'BUCKETSIZE': 1540}}, '*2': {'TYPE': 'IncrementalStMan', 'NAME': 'ism1', 'COLUMNS': ['coli2'], 'SPEC': {'CACHESIZE': 1, 'BUCKETSIZE': 32768}}, '*3': {'TYPE': 'TiledShapeStMan', 'NAME': 'tsm1', 'COLUMNS': ['colarrtsm'], 'SPEC': {'SEQNR': 2, 'HYPERCUBES': {}, 'DEFAULTTILESHAPE': array([], dtype=int32), 'MAXIMUMCACHESIZE': 0}}}
{'comment': '', 'dataManagerType': 'TiledShapeStMan', '_c_order': True, 'option': 0, 'valueType': 'dcomplex', 'maxlen': 0, 'dataManagerGroup': 'tsm1', 'ndim': 2}
{'TYPE': 'TiledShapeStMan', 'NAME': 'tsm2', 'SPEC': {'SEQNR': 3, 'HYPERCUBES': {}, 'DEFAULTTILESHAPE': array([], dtype=int32), 'MAXIMUMCACHESIZE': 0}}
[[[  1.+0.j   2.+0.j   3.+0.j]
  [  4.+0.j   5.+0.j   6.+0.j]]

 [[ 11.+0.j  12.+0.j  13.+0.j]
  [ 14.+0.j  15.+0.j  16.+0.j]]]
[[ 5.+0.j  6.+0.j]]
{'key1': 'keyval', 'keyrec': {'skey1': 1, 'skey2': 3.0}}
keyval
['key1', 'keyrec']
['key1', 'keyrec']
['skey1', 'skey2']
{}
['coli', 'cold', 'colarrtsm']
{'colarrtsm': array([[ 1.+0.j,  2.+0.j,  3.+0.j],
       [ 4.+0.j,  5.+0.j,  6.+0.j]]), 'cold': 4.0, 'coli': 1}
{'cold': 14.0, 'colarr': array([[ 1.+0.j,  2.+0.j,  3.+0.j],
       [ 4.+0.j,  5.+0.j,  6.+0.j]]), 'coli': 10}
[100000.]
[11  3]
[1 1] [2, 3]
[2 2 2] [1, 4, 5]
[3 3] [6, 7]
[4 4] [8, 9]
[5 5] [10, 11]
[6 6] [12, 13]
[7 7] [14, 15]
[8 8] [16, 17]
[9 9] [18, 19]
[10 10 10] [0, 20, 21]
[9, 8, 7, 6, 5, 4, 23]
[10, 2, 1, 1, 2, 2, 23, 3, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8, 9, 9, 10, 10]
False ['coli']
[6]
[]
[1, 4, 5]
[1, 4, 5, 7, 8, 9, 10, 11, 12, 13, 14, 15]
[7, 8, 9, 10, 11, 12, 13]
[1, 4, 5, 7, 8, 9, 10, 11, 12, 13]
